adding code and readme

burdinskid13 opened the pull request:
Adding code necessary to run preprocessing and analysis for the paper in its current state. Adding README to describe how the project directory should be set up to recreate this project and how the code should be executed.

burdinskid13 committed:
commiting code written for preprocessing DICOMs

burdinskid13 committed:
Delete .gitignore

Using local .gitignore instead for personal files

burdinskid13 committed:
Merge remote-tracking branch 'origin/main'

burdinskid13 committed:
adding two missing scripts to preprocessing (additional MID events population file and mriqc tsv creating file)

burdinskid13 committed:
committing analysis scripts for GLM analysis of the 3 tasks and behavioral analysis of the nback task

burdinskid13 committed:
updating environment yml file to reflect latest python environment

burdinskid13 committed:
removing older environment yml version

burdinskid13 committed:
Update README.md to include directory structure and code guidelines

satra commented on README.md:
```diff
@@ -1 +1,66 @@
+
+### Analysis:
+
+1. Demographics and cannabis characteristics by group table generation
+
+2. cifti smoothing using hcp tools 
+
+3. first, second (for mid and sst tasks given they have two runs) and group level modeling for nifti (volumetric) data
+
+4. first, second (for mid and sst tasks given they have two runs) and group level modeling for cifti (grayordinate) data
+
+5. fsleyes visualization for nifti/volumetric results
+
+6. final visualizations using fsleyes outputs and pillow for nifti/volumetric results
+
+7. final visualizations using nilearn and pillow for cifti/grayordinate results
```
May want to connect this to the relevant analysis scripts or sections

burdinskid13 replied:
```suggestion
1. Demographics and cannabis characteristics by group table generation

* Run the following jupyter notebooks in order:
  * _analysis/behavioral_analysis/cannabis_table.ipynb_
  * _analysis/behavioral_analysis/demographics_table.ipynb_


2. cifti smoothing using hcp tools 

* Run the following using SLURM job scheduler:
  * _analysis/HCP_smoothing/run_HCP_smoothing_for_tasks.sh_
    * this will call for each task: _analysis/HCP_smoothing/HCP_smoothing_by_tasks.sh_


3. first, second (for mid and sst tasks given they have two runs) and group level modeling for nifti (volumetric) data

* First level: Run the following using SLURM job scheduler:
  * Replace {task} with the task that you want to run the first level model for
  * _analysis/task_analysis_volume/submit_job_array_first_level.sh {task}_
    * this will call for each participant: _analysis/task_analysis_volume/first_level_analysis_ss.sh_
    * which will call for each participant: _analysis/task_analysis_volume/first_level_analysis_ss.py_

* Second level (only for mid and sst tasks given they have two runs): Run the following using SLURM job scheduler:
  * Replace {task} with the task that you want to run the first level model for
  * _analysis/task_analysis_volume/submit_job_array_second_level_vol.sh {task}_
    * this will call for each participant: _analysis/task_analysis_volume/second_level_analysis_vol_ss.sh_
    * which will call for each participant: _analysis/task_analysis_volume/second_level_analysis_vol_ss.py_

* Group level: Run the following jupyter notebook in order:
  * Set _task=_ in the block that loops through all relevant participants to the one you want to analyze
  * Note that you can use commenting to make modeling decisions, eg. outlier selection, sub-group analysis and covariates 
  * _analysis/task_analysis_volume/group_level_analysis_vol.ipynb_
    * this will save the effect size and stat maps per group/session/task at _../../../derivatives/task_analysis_volume/group_level/group-{group}/ses-{ses}/task-{task}_ to be used by the fsleyes visualization


4. first, second (for mid and sst tasks given they have two runs) and group level modeling for cifti (grayordinate) data

* First level: Run the following using SLURM job scheduler:
  * Replace {task} with the task that you want to run the first level model for
  * _analysis/task_analysis_surface/submit_job_array_first_level.sh {task}_
    * this will call for each participant: _analysis/task_analysis_surface/first_level_analysis_ss.sh_
    * which will call for each participant: _analysis/task_analysis_surface/first_level_analysis_ss.py_

* Second level (only for mid and sst tasks given they have two runs): Run the following using SLURM job scheduler:
  * Replace {task} with the task that you want to run the first level model for
  * _analysis/task_analysis_surface/submit_job_array_second_level_surf.sh {task}_
    * this will call for each participant: _analysis/task_analysis_surface/second_level_analysis_surf_ss.sh_
    * which will call for each participant: _analysis/task_analysis_surface/second_level_analysis_surf_ss.py_

* Group level: Run the following jupyter notebook in order:
  * Set _task=_ in the block that loops through all relevant participants to the one you want to analyze
  * Note that you can use commenting to make modeling decisions, eg. outlier selection, sub-group analysis and covariates 
  * _analysis/task_analysis_surface/group_level_analysis_surf.ipynb_
      * this will save the left/right hemispheres and flat maps as well as the coronal display per group/session/task at _../../../derivatives/task_analysis_surface/visualization/raw_figures_ to be used by the final pillow visualization


5. fsleyes visualization for nifti/volumetric results

* Run the following jupyter notebook in order:
  * Set _task=_ in the block that loops through all relevant participants to the one you want to analyze
  * _analysis/fsleyes_vis/fsleyes_visualization.ipynb_
    * this will save the fsleyes visualizations per group/session/task at _../../../derivatives/task_analysis_volume/visualization/raw_figures_ to be used by the final pillow visualization


6. final visualizations using fsleyes outputs and pillow for nifti/volumetric results

* Run the following jupyter notebook in order:
  * Set _task=_ in the block that loops through all relevant participants to the one you want to analyze
  * _analysis/task_analysis_volume/final_volume_plots.ipynb_
    * this will save the final visualizations per task at _../../../derivatives/task_analysis_volume/visualization/final_nifti_figures_ 


7. final visualizations using nilearn and pillow for cifti/grayordinate results

* Run the following jupyter notebook in order:
  * Set _task=_ in the block that loops through all relevant participants to the one you want to analyze
  * _analysis/task_analysis_surface/final_volume_plots.ipynb_
    * this will save the intermediate and final visualizations per task at _../../../derivatives/task_analysis_surface/visualization/intermediate_cifti_figures_ 
    * this will save the final visualizations per task at _../../../derivatives/task_analysis_surface/visualization/final_cifti_figures_ 


```

satra commented on analysis/HCP_smoothing/HCP_smoothing_by_task.sh:
```diff
@@ -0,0 +1,64 @@
+#SBATCH -x node[054-060,100-115]
+
+
+module load openmind/hcp-workbench/1.2.3
```
create a file in which you list all the requirements for the analysis. ideally this should be a reproducible environment. for now, just a listing of top level requirements would be a good starting point.

satra replied:
you can add this to the readme to complement your environment.yml file

burdinskid13 replied:
just did this!

satra commented on analysis/HCP_smoothing/HCP_smoothing_by_task.sh:
```diff
@@ -0,0 +1,64 @@
+        #command inputs in order: cifti file; surface kernel size in terms of sigma, volume kernel size in terms of sigma; direction to smooth along: COLUMN is used for dtseries files, while ROW would be used for dconn files; cifti output file; left surface file; right surface file
+        wb_command -cifti-smoothing \
+        $input_cifti_path \
+        4 4 \
```
if this is sigma as written in the comments above, then FWHM would be over 9.2mm, which seems large for the resolution in this study.

satra replied:
especially large for the volumetric structures.

burdinskid13 replied:
that's a good point. I was using 4 4 because Jeff and Gasser were using it (I think we all found an online example using this setting and just kept using it). for the subcortical volumetric structures I will change this to fwhm=6 (that's what i use in the volumetric analysis). what do you suggest for the vertex/surface data?

burdinskid13 replied:
added -fwhm to cifti-smoothing command

burdinskid13 replied:
```suggestion
       -fwhm 4 4 \
```

satra commented on analysis/behavioral_analysis/cannabis_table.ipynb:
```diff
@@ -0,0 +1,228 @@
+   "source": [
+    "def rm_CUD_baseline(subs_all):\n",
+    "    \n",
+    "    #subs to be excluded (only MM)\n",
```
perhaps add a comment as to why subs are excluded?

burdinskid13 replied:
```suggestion
    "    #subs to be excluded (only MM)\n",
    "    #these are excluded from the study because they had cannabis use disorder at baseline (exclusion criterium) 
```

burdinskid13 replied:
```suggestion
    "    #subs to be excluded (only MM) because they had cannabis use disorder at baseline (exclusion criterium)\n",
```

satra commented on analysis/behavioral_analysis/cannabis_table.ipynb:
```diff
@@ -0,0 +1,228 @@
+    "        os.makedirs (nilearn_output_path)\n",
+    "    \n",
+    "    if include_pval:\n",
+    "        mytable.to_csv('../../../derivatives/demographics/cannabis_table1_all_pval.csv')                          \n",
+    "        mytable_df = pd.read_csv('../../../derivatives/demographics/cannabis_table1_all_pval.csv')\n",
+    "        mytable_df.drop('Grouped by ses', axis=1, inplace=True)    #remove missing\n",
+    "        mytable_df.drop('Grouped by ses.1', axis=1, inplace=True)  #remove overall\n",
+    "        mytable_df = mytable_df[['Unnamed: 0','Unnamed: 1','Grouped by ses.3','Grouped by ses.2','Grouped by ses.4']] #swap 1 year and baseline columns\n",
+    "        mytable_df.to_csv('../../../derivatives/demographics/cannabis_table1_all_pval.csv',index=False)\n",
```
seems a little counter to save and then immediately overwrite. why save in the first place.

burdinskid13 replied:
because the format of mytable is not a dataframe that i can edit (i think it's a class in the table 1 package), so i the only way i was able to extract a pandas df from it was to save it as csv and then read it as a pandas df. then i edit and save again. i agree, it's not the nicest, but it was the most practical for what i wanted to do

satra commented on analysis/behavioral_analysis/cannabis_table.ipynb:
```diff
@@ -0,0 +1,228 @@
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "include_pval=True\n",
+    "create_table(include_pval)"
```
for a notebook like this, unless that function gets reused, it's ok to simply leave it as a script.

burdinskid13 replied:
i'm not sure what you mean here? i print it out without p-values for the paper but i also checked the p-values once beforehand (so i use both True and False)

satra commented on analysis/behavioral_analysis/demographics_table.ipynb:
```diff
@@ -0,0 +1,218 @@
+    "        os.makedirs (nilearn_output_path)\n",
+    "    \n",
+    "    if include_pval:\n",
+    "        mytable.to_csv('../../../derivatives/demographics/table1_all_pval.csv')\n",
+    "        mytable_df = pd.read_csv(f'../../../derivatives/demographics/table1_all_pval.csv')\n",
+    "        mytable_df.drop('Grouped by group_ses.1', axis=1, inplace=True)     \n",
+    "        mytable_df.to_csv('../../../derivatives/demographics/table1_all_pval.csv',index=False)\n",
```
same pattern here as before.

satra replied:
unless TableOne doesn't return a pandas dataframe or cannot be converted to one.

satra commented on analysis/behavioral_analysis/non_imaging_analysis_nback_group_comparison.ipynb:
```diff
@@ -0,0 +1,549 @@
+    "            dict_MM_baseline_cudit = non_img_data[non_img_data['SSS.CHR.Time_point'] == 'Baseline'].groupby('IDS.CHR.Subject')['INV.INT.CUDIT.Summed_score'].agg(\"first\").to_dict()\n",
+    "            df_subs['total_cudit'] = df_subs['subs'].map(dict_MM_baseline_cudit)\n",
+    "            df_subs['total_cudit'] = df_subs['total_cudit'] - df_subs['total_cudit'].mean()\n",
+    "            df_subs['total_cudit'].fillna(0, inplace=True)\n",
```
this does mean that you are saying the people missing have the same as the mean cudit. in general, this should either be imputed or the subjects with missing data removed.

satra replied:
one way to remove these subjects is to add columns with a 1 for each subject with a missing value. similar to how at the first level outlier scans are omitted.

satra commented on analysis/behavioral_analysis/non_imaging_analysis_nback_group_comparison.ipynb:
```diff
@@ -0,0 +1,549 @@
+    "\n",
+    "    for column in nback_data.columns[1:]:\n",
+    "        nback_data[column] = nback_data[column] - nback_data[column].mean()\n",
+    "        nback_data[column].fillna(0, inplace=True)\n",
```
same comment as before.

satra commented on analysis/task_analysis_surface/first_level_analysis_ss.py:
```diff
@@ -0,0 +1,365 @@
+                         ]
+    
+    #for loop that creates complex contrasts based on the string names from the task_contrasts list created above
+    for task_contrast in task_contrasts:
+        #split complex contrast string into a list of canonical contrasts separated by '-' or '+' 
```
this seems like a complex section of code. perhaps just spot check that the vectors are correct

satra commented on analysis/task_analysis_surface/first_level_analysis_ss.py:
```diff
@@ -0,0 +1,365 @@
+    
+    #save old header info
+    orig_cifti_nimg = nimg.load_img(orig_cifti)
+    time_axis, brain_model_axis = [orig_cifti_nimg.header.get_axis(i) for i in range(orig_cifti_nimg.ndim)]
```
does this do what you expect it to do, when ndim is more than 2 ? there may be a cleaner way to do this.

burdinskid13 replied:
added commit that catches incorrect number of axes

satra commented on analysis/task_analysis_surface/first_level_analysis_ss.py:
```diff
@@ -0,0 +1,365 @@
+        #save contrast maps to files
+        dscalar_output['effect_size'].to_filename(f'../../../derivatives/task_analysis_surface/first_level/sub-{sub}/ses-{ses}/task-{task}/sub-{sub}_ses-{ses}_task-{task}_rec-unco_run-{run}_contrast-{contrast}_effect_size.dscalar.nii')
+        dscalar_output['effect_variance'].to_filename(f'../../../derivatives/task_analysis_surface/first_level/sub-{sub}/ses-{ses}/task-{task}/sub-{sub}_ses-{ses}_task-{task}_rec-unco_run-{run}_contrast-{contrast}_effect_variance.dscalar.nii')
+        dscalar_output['z_score'].to_filename(f'../../../derivatives/task_analysis_surface/first_level/sub-{sub}/ses-{ses}/task-{task}/sub-{sub}_ses-{ses}_task-{task}_rec-unco_run-{run}_contrast-{contrast}_z_score.dscalar.nii')
+        dscalar_output['stat'].to_filename(f'../../../derivatives/task_analysis_surface/first_level/sub-{sub}/ses-{ses}/task-{task}/sub-{sub}_ses-{ses}_task-{task}_rec-unco_run-{run}_contrast-{contrast}_stat.dscalar.nii')
+        dscalar_output['p_value'].to_filename(f'../../../derivatives/task_analysis_surface/first_level/sub-{sub}/ses-{ses}/task-{task}/sub-{sub}_ses-{ses}_task-{task}_rec-unco_run-{run}_contrast-{contrast}_p_value.dscalar.nii')
```
just to clarify, stat gets converted to pvalue and that gets converted to zscore by nilearn. so all three are effectively stats.

burdinskid13 replied:
yea thank you, that lab meeting discussion was really helpful! i was just maintaining the nilearn convention for naming

satra commented on analysis/task_analysis_surface/group_level_analysis_surf.ipynb:
```diff
@@ -0,0 +1,994 @@
+    "def rm_CUD_baseline(effect_size_maps, all_subs):\n",
+    "    \n",
+    "    #subs to be excluded (only MM)\n",
+    "    excluded_subs = ['MM014','MM188','MM197','MM217','MM228','MM239','MM241']\n",
```
a comment here would be useful why these were excluded.

burdinskid13 replied:
added! and did same for volumetric group analysis

satra commented on analysis/task_analysis_surface/group_level_analysis_surf.ipynb:
```diff
@@ -0,0 +1,994 @@
+    "                group_design_matrix['total_cudit'] = df_subs['subs'].map(dict_MM_baseline_cudit)\n",
+    "                group_design_matrix['total_cudit'] = group_design_matrix['total_cudit'] - group_design_matrix['total_cudit'].mean()\n",
+    "                #needed because MM_141 is nan, so replacing with 0, which is the mean\n",
+    "                group_design_matrix['total_cudit'].fillna(0, inplace=True)\n",
```
missing data, also make sure subjects are removed in the corresponding analysis.

burdinskid13 replied:
I was doing imputation by the mean -- it's not the best but it's not a major part of the analysis and i thought it would be ok. what do you suggest instead? you think i should remove the subject?

(same comment as for other area where this happened)

satra replied:
take a look at figure 1 in this paper (https://elifesciences.org/articles/54055) for an approach.

in terms of this paper, i think you can go ahead with the submission, but it would be good to note how many people are missing data in those columns. 

it would be good to add supplemental analysis to see what happens when you leave people out or impute using other data (not just the mean of the same feature) or use the method in the elife paper.

satra commented on analysis/task_analysis_surface/second_level_analysis_surf_ss.py:
```diff
@@ -0,0 +1,149 @@
+        
+        #TO DO: figure out meaning of precision weighted
+        #note that results are made up of these: fx_results = [fixed_fx_contrast, fixed_fx_variance, fixed_fx_tstat]
+        fx_results = _compute_fixed_effects_params(
```
does nilearn really require a hidden function to compute this? is there no public api for this?

satra replied:
or is this because of cifti?

burdinskid13 replied:
yes it's because of cifti. that's the only time when i went to hidden functions

satra commented on analysis/task_analysis_surface/second_level_analysis_surf_ss.py:
```diff
@@ -0,0 +1,149 @@
+        
+        
+        
+def create_dscalar(brain_model_axis, contrast_output):
```
some of these functions could be put into a single python utils library and used across the scripts. not essential for submission, but minimizes errors.

burdinskid13 replied:
I did not know about this option, but good to know for next time! i agree, i did not like having to copy the same function into several scripts

satra commented on analysis/task_analysis_volume/final_volume_plots.ipynb:
```diff
@@ -0,0 +1,217 @@
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "def create_vol_figure_final(task,contrasts,w_no_cmap):\n",
```
for the first paper i was involved with as a grad student, we created final figure and then went on to produce revisions of that final figure up to 8 times i think. since then i don't call anything final.

burdinskid13 replied:
renamed for surf and vol!

satra commented on analysis/task_analysis_volume/group_level_analysis_vol.ipynb:
```diff
@@ -0,0 +1,1013 @@
+    "                group_design_matrix['total_cudit'] = df_subs['subs'].map(dict_MM_baseline_cudit)\n",
+    "                group_design_matrix['total_cudit'] = group_design_matrix['total_cudit'] - group_design_matrix['total_cudit'].mean()\n",
+    "                #needed because MM_141 is nan, so replacing with 0, which is the mean\n",
+    "                group_design_matrix['total_cudit'].fillna(0, inplace=True)\n",
```
missing issue. check.

burdinskid13 replied:
I was doing imputation by the mean -- it's not the best but it's not a major part of the analysis and i thought it would be ok. what do you suggest instead? you think i should remove the subject?

(same comment as for other area where this happened)

satra commented on analysis/task_analysis_volume/group_level_analysis_vol.ipynb:
```diff
@@ -0,0 +1,1013 @@
+    "                    stat_map = contrast_output['z_score']\n",
+    "                    effect_map = contrast_output['effect_size']\n",
+    "\n",
+    "                    #for \"positive\" activation difference (nicer visualization)\n",
```
if this is used, make sure that the figure description says so. why not just use the reverse contrast definition in the first place?

burdinskid13 replied:
it was a quick fix to not have to rerun things. do you think it's worth editing it all?

satra replied:
aren't group contrasts easy to compute? you could flip the contrast at the group level, no?

satra replied:
this is more a notion of the kind of thing that gets lost in details. I don't think you need to change the code at the moment, but make sure that in the paper wherever this contrast is mentioned or shown, the appropriate contrast is described.

satra replied:
also make a note in the place where the contrast is specified in the script that the plots reverse the effect.

satra reviewed (commented):
left a bunch of comments in various places. the main one i would say that you should look into is the replacement of missing with 0.

you may not want to delete the `.gitignore` as it helps not add things like DS store into the git commit.

burdinskid13 committed:
Update analysis/behavioral_analysis/cannabis_table.ipynb

burdinskid13 committed:
Update cannabis_table.ipynb

restore previous version that was accidentally messed up when replying to a comment in my pull request

burdinskid13 committed:
Update analysis/behavioral_analysis/cannabis_table.ipynb

add explanation for why certain subs are excluded from the cannabis table

burdinskid13 committed:
Update README.md

adding specifics to analysis instructions

burdinskid13 committed:
Update README.md

changes to readme to update necessary coding environment and to include detailed steps to carry out preprocessing

burdinskid13 committed:
Update HCP_smoothing_by_task.sh

adding -fwhm flag to cifti-smoothing hcp workbench command

burdinskid13 committed:
Update first_level_analysis_ss.py

add error catch for incorrect number of axes in cifti header

burdinskid13 committed:
Update group_level_analysis_surf.ipynb

adding note about CUD at baseline MMC subject also in function that removes them

burdinskid13 committed:
Update group_level_analysis_surf.ipynb

remove additional unneeded line about subs to be excluded based on CUD at baseline

burdinskid13 committed:
Update group_level_analysis_vol.ipynb

add note to function where subs are removed based on CUD at baseline why they are removed

burdinskid13 committed:
Update and rename final_volume_plots.ipynb to figures_vol.ipynb

rename function call avoiding the word "final"

burdinskid13 committed:
Update and rename final_surface_plots.ipynb to figures_surf.ipynb

avoid using the word final in function name

burdinskid13 committed:
preprocessing and analysis code as of Feb 22, 2024

burdinskid13 committed:
Update README.md to reflect code changes saved on Feb 22, 2024

burdinskid13 committed:
added change score model and changed freq encoding

burdinskid13 committed:
updated readme

burdinskid13 committed:
analysis code changes for second submission of study

burdinskid13 committed:
updating README.md accordingly with the second submission

burdinskid13 committed:
changing IQR representation to range format in table 1 and etable1
